backend/de.metas.business/src/main/java/de/metas/fulltextsearch/indexer/handler/invoice/InvoiceFTSModelIndexer.java

```diff
@@ -0,0 +1,243 @@
+		if (invoiceIds.isEmpty())
+		{
+			return InvoiceFTSModelIndexer.ChangedDocumentIds.EMPTY;
+		}
+
+		final String sql = "SELECT * from C_Invoice_Adv_Search_Update(p_C_Invoice_IDs := ARRAY["
+				+ invoiceIds.stream().map(invoiceId -> "?").collect(Collectors.joining(","))
+				+ "])";
+		final ArrayList<Object> sqlParams = new ArrayList<>(invoiceIds);
+
+		final InvoiceFTSModelIndexer.ChangedDocumentIds.ChangedDocumentIdsBuilder changes = InvoiceFTSModelIndexer.ChangedDocumentIds.builder();
+
+		DB.forEachRow(sql, sqlParams, rs -> {
+			final String esDocumentId = rs.getString("es_documentid");
+			final String operation = rs.getString("operation");
+			if ("D".equals(operation))
+			{
+				changes.removedDocumentId(esDocumentId);
+			}
+			else
+			{
+				changes.changedDocumentId(esDocumentId);
+			}
+		});
+
+		return changes.build();
+	}
+
+	private ImmutableSet<InvoiceId> extractAffectedInvoiceIds(@NonNull final List<ModelToIndex> requests)
+	{
+		final ImmutableSet.Builder<InvoiceId> invoiceRepoIds = ImmutableSet.builder();
+		final HashSet<BPartnerId> bpartnerRepoIds = new HashSet<>();
+		final HashSet<Integer> bpartnerLocationRepoIds = new HashSet<>();
+		final HashSet<Integer> bpartnerContactRepoIds = new HashSet<>();
+		final HashSet<DocTypeId> docTypeRepoIds = new HashSet<>();
+		final HashSet<WarehouseId> warehouseRepoIds = new HashSet<>();
+		final HashSet<CalendarId> calendarRepoIds = new HashSet<>();
+		final HashSet<YearId> yearRepoIds = new HashSet<>();
+
+		for (final ModelToIndex request : requests)
+		{
+			final TableRecordReference sourceModelRef = request.getSourceModelRef();
+			if (sourceModelRef == null)
+			{
+				continue;
+			}
+
+			final String sourceTableName = sourceModelRef.getTableName();
+			if (I_C_Invoice.Table_Name.equals(sourceTableName))
+			{
+				invoiceRepoIds.add(InvoiceId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_BPartner.Table_Name.equals(sourceTableName))
+			{
+				bpartnerRepoIds.add(BPartnerId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_BPartner_Location.Table_Name.equals(sourceTableName))
+			{
+				bpartnerLocationRepoIds.add(sourceModelRef.getRecord_ID());
+			}
+			else if (I_AD_User.Table_Name.equals(sourceTableName))
+			{
+				bpartnerContactRepoIds.add(sourceModelRef.getRecord_ID());
+			}
+			else if (I_C_DocType.Table_Name.equals(sourceTableName))
+			{
+				docTypeRepoIds.add(DocTypeId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_M_Warehouse.Table_Name.equals(sourceTableName))
+			{
+				warehouseRepoIds.add(WarehouseId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_Calendar.Table_Name.equals(sourceTableName))
+			{
+				calendarRepoIds.add(CalendarId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_Year.Table_Name.equals(sourceTableName))
+			{
+				yearRepoIds.add(YearId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+		}
+		invoiceDAO.retrieveInvoiceIdsByBPartnerIds(bpartnerRepoIds)
```

Comment:
pls join all those retrieves in one "query that does all",
(conceptually something like `C_BPartner_ID IN (...) OR C_BP_Location_ID IN (...) OR AD_User_ID IN (...)` etc)